mindsdb/integrations/handlers/llama_index_handler/llama_index_handler.py

```diff
@@ -0,0 +1,108 @@
+import os
+from typing import Optional, Dict
+
+from llama_index import GPTVectorStoreIndex,download_loader
+import pandas as pd
+
+from mindsdb.utilities.config import Config
+from mindsdb.integrations.libs.base import BaseMLEngine
+
+from mindsdb.utilities.log import get_log
+
+
+logger = get_log("integrations.llama_index_handler")
+
+
+
+class LlamaIndexHandler(BaseMLEngine):
+    """
+    Integration with the LlamaIndex Python Library
+    """
+    name = 'llama_index'
+
+    def __init__(self, *args, **kwargs):
+
+        super().__init__(*args, **kwargs)
+        self.default_index_class  = 'GPTVectorStoreIndex'
+        self.default_query_engine = 'as_query_engine' 
+        self.supported_index_class = ['GPTVectorStoreIndex']
+        self.supported_query_engine = ['as_query_engine'] 
+
+
+    def create(self, target: str, df: Optional[pd.DataFrame] = None, args: Optional[Dict] = None) -> None:
+
+        if 'using' not in args:
+            raise Exception("LlamaIndex engine requires a USING clause! Refer to its documentation for more details.")
+
+        if 'source_url_link' not in args['using']:
+            raise Exception("LlamaIndex engine requires a source_url_link parameter.Refer to its documentation for more details.")
+
+        if 'openai_api_key' not in args['using']:
+            raise Exception("LlamaIndex engine requires a openai_api_key parameter.Refer to its documentation for more details.")
+
+        if 'index_class' not in args['using']:
+            args['using']['index_class'] = self.default_index_class
+        elif args['using']['index_class'] not in self.supported_index_class:
+            raise Exception(f"Invalid index class argument. Please use one of {self.supported_index_class}")
+
+        if 'query_engine' not in args['using']:
+            args['using']['query_engine'] = self.default_query_engine
+        elif args['using']['query_engine'] not in self.supported_query_engine:
+            raise Exception(f"Invalid operation mode. Please use one of {self.supported_query_engine}")
+
+        self.model_storage.json_set('args', args)
+
+
+    def predict(self, df: Optional[pd.DataFrame] = None, args: Optional[Dict] = None) -> None:
+        
+        args = self.model_storage.json_get('args')
+
+        input_keys = list(args.keys())
+
+        os.environ['OPENAI_API_KEY'] = args['using']['openai_api_key']
+
+        SimpleWebPageReader = download_loader("SimpleWebPageReader")
+        documents = SimpleWebPageReader(html_to_text=True).load_data([args['using']['source_url_link']])
```

Comment:
curious: how come we need to reload this upon each predict call? If I understand correctly, the predict calls only change the questions, but the context remain fixed.
I'd imagine we can build the index once and serve any number of predict calls thereafter.

Reply:
I agree; I will address this change in sometime

Reply:
Addressed the changes. Please kindly check this